Add compiler warning tracking tooling failure remediation guidance (#1364)
python/devguide: development-tools/warnings.rst

.. warnings:

Tools for tracking compiler warnings
====================================

.. highlight:: bash

The compiler warning tracking tooling is intended to alert developers about new
compiler warnings introduced by their contributions. The tooling consists of
a Python script which is ran by the following GitHub workflows:

* Ubuntu/build and test (:cpy-file:`.github/workflows/reusable-ubuntu.yml`)
* macOS/build and test (:cpy-file:`.github/workflows/reusable-macos.yml`)

> You can check the documentation for the :cpy-file:`Tools/build/check_warnings.py` tool
> by running::
>    python Tools/build/check_warnings.py --help

Reviewer: ```suggestion
You can check the documentation for the :cpy-file:`Tools/build/check_warnings.py` tool
by running::

   python Tools/build/check_warnings.py --help
```

Author: Added newline

.. warnings:

> =================================

Reviewer 2: We can skip this overline:
```suggestion
```

The convention is here: https://devguide.python.org/documentation/markup/#sections

We often skip the overlines though.

Author: Removed overfine. I had just copied how other pages in development-tools were under/overlining.

> Compiler Warning Tracking Tooling

Reviewer 3: This title feels a bit buzzword-y -- perhaps, though more boring:
```suggestion
Tools for tracking compiler warnings
```

> Compiler Warning Tracking Tooling

Author: Changed to "Tools for tracking compiler warnings". Will keep it sentence case in the future.

> The compiler warning tracking tooling is intended to alert developers to new

Reviewer 2: Sure, though we only two code-blocks with different highlights :)

> The compiler warning tracking tooling is intended to alert developers to new

Author: Added the default anyway!

> The compiler warning tracking tooling is intended to alert developers to new

Author: changed to about

compiler warnings introduced by their contributions. The tooling consists of

> Python script which is ran by a few GitHub Actions workflows. These

Author: Added a

> python script which is ran by a few GitHub Actions worksflows. Currently these

Author: Capitalized and made the following sentence more direct

> GitHub Actions jobs run the warning checking tooling:

Author: Applied suggestion to make this more concise.

- Ubuntu/Build and Test

> - macOS/Build and Test

Author: Included links to the configuration files

- macOS/Build and Test

The help text for the `Tools/build/check_warnings.py` is as follows:

> ::

Reviewer 3: An alternative to `none` is simply [`text`](https://pygments.org/docs/lexers/#pygments.lexers.special.TextLexer). The [full list of languages](https://pygments.org/docs/lexers) comes from Pygments.

> ::

Author: Nice I like how it links to the file in the repo! Changed to a link and made the code block format "text"

    usage: check_warnings.py [-h]
           -c COMPILER_OUTPUT_FILE_PATH
           [-i WARNING_IGNORE_FILE_PATH] [-x] [-X] -t {json,clang}

    options:
        -h, --help            show this help message and exit
        -c COMPILER_OUTPUT_FILE_PATH, --compiler-output-file-path COMPILER_OUTPUT_FILE_PATH
                                                    Path to the compiler output file
        -i WARNING_IGNORE_FILE_PATH, --warning-ignore-file-path WARNING_IGNORE_FILE_PATH
                                                    Path to the warning ignore file
        -x, --fail-on-regression
                                                    Flag to fail if new warnings are found
        -X, --fail-on-improvement
                                                    Flag to fail if files that were expected to have warnings have no warnings
        -t {json,clang}, --compiler-output-type {json,clang}

>                                                     Type of compiler output file (json or clang)

Author: Probably a good idea. I figured this was a quick way for the reader to get an intuition of what inputs are going to the script if their only exposure to the tooling is that a CI failure has sent them here. I removed it to future-proof this guide.

The script can be run locally by providing the compiler output and the
compiler output type to see a list of unique warnings.

::

        $ python Tools/build/check_warnings.py -c <compiler_output_file> -t <compiler_output_type>

..

    Note: `-fdiagnostics-format=json` flag is required when compiling with GCC
           for the script to properly parse the compiler output.

.. _warning-check-failure:

> What To Do If Warning Check Fails GitHub CI
> -------------------------------------------

Reviewer 2: ```suggestion
What to do if a warning check fails GitHub CI
---------------------------------------------
```

Author: Made sentence case

The warning check tooling will fail if the compiler generates more or less
warnings than expected for a given source file as defined in the platform

> specific warning ignore file. The warning ignore file is either

Author: hyphenated platform-specific

:cpy-file:`Tools/build/.warningignore_ubuntu` or :cpy-file:`Tools/build/.warningignore_macos`

> depending on the platform.

Author: Is `:cpy-fil:`!check_warnings.py` supposed to be shorthand to link to a file with the same basename already referenced in the document? When build the dev guide with the suggestion it would take me to `cpython/check_warnings.py`.

I put in the full path for now.

> depending on the platform.

Reviewer: The `!` tells Sphinx not to generate a link, since for brevity I only used the filename and not the full path, a link can't be generated.  By still using `:cpy-file:` the filename is semantically marked and has the same style as the other filenames marked with the same role.  The file is already linked above, so there is no need to repeat the link.

Note that an alternative would have been `` :cpy-file:`check_warnings.py <Tools/build/check_warnings.py>` ``, which would have linked to the correct path while only showing the filename, but that doesn't seem to be supported by the `:cpy-file:` role.

> depending on the platform.

Author: ```
:cpy-file:`!check_warnings.py`
```

Renders as

<img width="309" alt="Screenshot 2024-08-10 at 7 34 43 PM" src="https://github.com/user-attachments/assets/d7a03d48-1de0-4c98-84a2-13ea8a55ef99">

And actually links to `cpython/check_warnings`. Is that what you are suggesting?

> If warning check fails with:

Author: added a

> * Unexpected Warnings

Reviewer 2: ```suggestion
* Unexpected warnings
```

Author: fixed capitalization

> * Unexpected Warnings

Reviewer: ```suggestion
* More warnings than expected:
```

Author: I kept the language the same but fixed the capitalization.

The reason being that the script actually labels warnings as "Unexpected improvements" and "Unexpected warnings"

    * Attempt to refactor the code to avoid the warning.
    * If it is not possible to avoid the warning document in the PR why it is

>        reasonable to ignore and add the warning to the platform specific

Reviewer 2: ```suggestion
       reasonable to ignore and add the warning to the platform-specific
```

Author: Hyphenated `platform-specific`

       warning ignore file. If the file exists in the warning ignore file

>        increment the count by the number of new introduced warnings.

Reviewer 2: ```suggestion
       increment the count by the number of newly introduced warnings.
```

Author: changed to newly

>     * If it is not possible to avoid the warning document in the PR why it is
>        reasonable to ignore and add the warning to the platform specific
>        warning ignore file. If the file exists in the warning ignore file
>        increment the count by the number of new introduced warnings.

Reviewer: ```suggestion
    * If it is not possible to avoid the warning document in the PR why it is
      reasonable to ignore and add the warning to the platform-specific warning
      ignore file. If the file already exists in the warning ignore file
      increment the count by the number of newly introduced warnings.
```

Author: Fixed spacing

> * Unexpected Imrpovements (Less Warnings)

Reviewer 2: ```suggestion
* Unexpected improvements (less warnings)
```

Author: fixed capitalization

> * Unexpected Imrpovements (Less Warnings)

Reviewer: ```suggestion
* Less warnings than expected
```

Author: I kept the language the way it was but fixed the capitalization, same reason as above the script actually labels specific warnings as such.

    * Document in the PR that the change reduces the number of compiler

>        warnings. Decrement the count in the platform specific warning

Reviewer 2: ```suggestion
       warnings. Decrement the count in the platform specific-warning
```

Author: Hyphenated `platform-specific` since that is what I think you meant and one I missed from your last review!

> 1. Unexpected Warnings
>     a. Attempt to refactor the code to avoid the warning.
>     b. If it is not possible to avoid the warning document in the PR why it is
>        reasonable to ignore and add the warning to the platform specific
>        warning ignore file. If the file exists in the warning ignore file
>        increment the count by the number of new introduced warnings.
> 2. Unexpected Imrpovements (Less Warnings)
>     a. Document in the PR that the change reduces the number of compiler
>        warnings. Decrement the count in the platform specific warning
>        ignore file or remove the file if the count is now zero.

Reviewer 2: ```suggestion
* Unexpected warnings
    * Attempt to refactor the code to avoid the warning.
    * If it is not possible to avoid the warning, document in the PR why it is
      reasonable to ignore and add the warning to the platform-specific
      warning ignore file. If the file exists in the warning ignore file
      increment the count by the number of newly introduced warnings.
* Unexpected improvements (less warnings)
    * Document in the PR that the change reduces the number of compiler
      warnings. Decrement the count in the platform-specific warning
      ignore file or remove the file if the count is now zero.
```

Author: I wanted this to read more like a decision tree and it looked like the only way for me to create a numbered list in rst was to explicitly create one. I changed this a bulleted list as suggested for now because when we start getting concrete examples of how to address warnings I want to really expand on this section.

Reviewer 2: I see. I usually follow the Google guidelines on this; use bullets unless it's a sequence of steps to be performed in order: https://developers.google.com/style/lists

Author: Probably what I want is bullets and then the second level is a sequence of steps, since that is the case here. The top level is a condition: you failed the test because of one of these conditions and then what follows are the steps to address it.

>     * Document in the PR that the change reduces the number of compiler
>        warnings. Decrement the count in the platform specific warning
>        ignore file or remove the file if the count is now zero.

Reviewer: ```suggestion
    * Document in the PR that the change reduces the number of compiler
      warnings. Decrement the count in the platform-specific warning
      ignore file or remove the file if the count is now zero.
```

Author: Fixed spacing
